### ISSUE TITLE
Forecasted Grants

### ISSUE THREAD

**Issue by ClaireValdivia** [usdigitalresponse/usdr-gost#2201]
### Why is this issue important?
Partner request and key area where grant finder returns less results than grants.gov as we don't include Forecasted grants in results. This is important to partners because it can take a lot of time to build a team to apply for a grant and pull the application together, so having as much lead time as possible increases the chances of an applicant being able to apply and secure the grant.

### Current State
No forecasted grants are ingested to grant finder and therefore are not included in search

### Expected State
Issue [330](https://github.com/usdigitalresponse/grants-ingest/issues/330) exists for ingest 

[Figma designs](https://www.figma.com/file/N1wuvwvciNhdKZNSdomTqn/Forecasted?type=design&node-id=0-1&mode=design&t=0fgpLbStzMSOvNAH-0)

**Search Drawer**
- When creating a new search or editing an existing search, the field "Opportunity Status" should show an additional option for "Forecasted" that displays to the left of "Posted"
- For any new search, Forecasted should be defaulted to be checked

**Results Table - General**
- On the results table that shows on My Grants and Browse Grants tabs, in the "Status" column, "Forecasted" should appear for any grants with that status
- Other statuses should be updated so the first letter is capitalized for consistency (i.e. Closed, Archived, Posted, Forecasted)
- When you click "Export to CSV" from the grants table, in column E of the file, the relevant Opportunity status should display (i.e. Closed, Archived, Posted, Forecasted)
- estimated close dates should sort in with regular dates. so est. 5/5/25 would be next to est 5/5/25
- "see details" should sort above not yet posted but below all dates
- "all grants" view of results table should display posted + forecasted (not closed or archived)
- default sort of results table should display chronologically by created date with the most recently created displaying first (I'm not sure the field name for this as it's not a field that displays on grants.gov or grant finder currently)

**Results Table - Open Date and Close Date**
Note that this includes changes to more than the forecasted open/close dates, but we want to make the changes together to be consistent.
- If opp status = forecasted, then display the Estimated Post Date in the "Open Date" column. Anytime this field is updated in grants.gov, we should expect to see the change reflected in Grant Finder.
-  If opp status = forecasted, and Estimated Application Due Date has a date entered in grants.gov, then display "est." and the date from Estimated Application Due Date field in the Close Date column of the results table (i.e. "est. 12/20/2023")
-  If opp status = forecasted and Estimated Application Due Date is blank in grants.gov - then display "Not yet issued" in the Close Date column
-  if opp status = Forecasted, and Estimated Application Due Date is anything other than a date or blank in grants.gov, display: "see details" in the Close Date column. "See details" should display like a link (blue text, underline on hover). Clicking the link (or anywhere in the whole row) should bring the user to the grant details page.
- When the opp status changes from Forecasted to Posted, the Posted Date should display in the "Open Date" column with the current value for that field, and the Close Date should display according to the rules below for posted grants
- If close date is blank AND opp status = posted - then display "Not yet issued" in the Close Date column
- if opp status = Posted, Archived, or Closed, and Close Date is anything other than a date or blank, display: "see details" in the Close Date column. "See details" should display like a link (blue text, underline on hover). Clicking the link (or anywhere in the whole row) should bring the user to the grant details page.

**Grant Details Page**
For any forecasted grants:
- The "Opportunity Status" field should display "Forecasted" for any forecasted grants
- Estimated Post Date should display in the "Open Date" field (i.e. "_est._ Jul 6, 2024")
- If Estimated Application Due Date has a date entered in grants.gov, then display "est. [date]" in the Close Date field (i.e. _est._ Sep 20, 2025)
- if Estimated Application Due Date is blank in grants.gov, display "Not yet issued" in the Close Date field
- If Estimated Application Due Date is anything other than a date or blank in grants.gov, display the full content of the field with text wrapping in the Close Date field
- When a grant changes from Forecasted to Posted, the Open Date field should display the Posted Date from grants.gov, and the Close Date field should display the Close Date from grants.gov

### Implementation Plan

Currently WIP

Search Drawer:
SearchPanel.vue -> add forecasted option to opportunityStatusOptions
SearchPanel.vue -> defaultCriteria -> add forecasted option to opportunityStatuses

Results Table - General:
grants.js -> criteria.opportunityStatuses = ['forecasted', 'posted'];
GrantsTable.vue → formattedGrants -> status: titleize(grant.opportunity_status)

**Comment by ClaireValdivia** [comment]
@caitlinwinner fyi that I scheduled a meeting next week to discuss this one with me, you, and Pari, but flagging for you here as well in case you have initial thoughts on the open questions listed above

**Comment by ClaireValdivia** [comment]
Next step on this issue is to add design mock-ups and update requirements on the issue accordingly

**Comment by caitlinwinner** [comment]
Designs : https://www.figma.com/file/N1wuvwvciNhdKZNSdomTqn/Forecasted?type=design&node-id=0%3A1&mode=design&t=XfCrk3DmM98qakJR-1

**Comment by ClaireValdivia** [comment]
acceptance criteria needs to be fleshed out and compared with #2066

**Comment by ClaireValdivia** [comment]
Need to add details on how 'est' dates will sort
Also confirm that "see details" dates shows at the end when sorted
*designs need to be updated to show current grant details page & ensure that requirements align

**Comment by caitlinwinner** [comment]
- estimated close dates should sort in with regular dates. so est. 5/5/25 would be next to est 5/5/25
- see details should sort above not yet posted but below all dates
- designs updated at the same link above

**Comment by masimons** [comment]
@ClaireValdivia regarding "For any new search, Forecasted should be defaulted to be checked" -> does this mean that we still only want to show posted grants on initial page load? We don't want to show forecasted grants when the table is first loaded, is this correct?

**Comment by ClaireValdivia** [comment]
@masimons good catch! i think it would make sense to have both posted and forecasted show for "all grants", but curious for thoughts on if this would impact the load time of the page significantly if there are more total grants to load?

**Comment by masimons** [comment]
@ClaireValdivia Since we are paginating, the load time _should_ be the same as it is now per page. There will just be more pages to click through, as there will be more grants. But as far as I can tell, there is currently a max of 50 records retrieved for the grants table per page.

**Comment by ClaireValdivia** [comment]
awesome thank you! I'll adjust the requirements to reflect:
- all grants should display posted + forecasted
- default sort should be to display chronologically by created date with the most recently created displaying first

**Comment by masimons** [comment]
@ClaireValdivia regarding "If opp status = forecasted and Estimated Application Due Date is blank in grants.gov - then display "Not yet issued" in the Close Date column" - shall we display "Not yet issued" for any grant that has a blank posted date, not just forecasted grants? Or do we actively prefer for posted grants without a close date to not display anything? (I imagine that the happy path for posted grants is that they always have a non-blank close date, but just trying to consider consistency for the edge cases)

**Comment by ClaireValdivia** [comment]
@masimons ah, I think it's addressed in a different bullet in that section: "If close date is blank AND opp status = posted - then display "Not yet issued" in the Close Date column" - does that address your question?

**Comment by masimons** [comment]
@ClaireValdivia yes! sorry - thanks for the quick response. I have a color-coded google doc to keep track of these specs, and I miscolored it!

**Comment by masimons** [comment]
https://github.com/usdigitalresponse/usdr-gost/pull/3346

Regarding "if opp status = Posted, Archived, or Closed, and Close Date is anything other than a date or blank, display: "see details" in the Close Date column" - from what I'm seeing, `close_date` in the database is a date type, which would mean that the close date would never be anything other than a date or blank. Is there an example of a grant we have on prod where we are seeing the close date as something other than blank or a date?

I did notice a field `close_date_explanation` on the grants table - would it make sense to display "See details" if:
- the `close_date` is blank, and `close_date_explanation` is not blank?

**Comment by ClaireValdivia** [comment]
Hmm this may be a case of me not knowing where things pull from in the table currently. This is an example of a grant where the close date is not a date: https://staging.grants.usdr.dev/grants/103313 -- is the text in the close date field pulling from `close_date_explanation`?

**Comment by masimons** [comment]
Hey @ClaireValdivia ! It looks like right now, on the grant details page, we are showing the `close_date_explanation` when we don't have a date - and if we also don't have any text for `close_date_explanation`, we show the default text "Not available".

Would we want to mirror this logic for the forecasted grants as well, on the grants detail page?

Separately though, for the grants table, I'm not seeing any logic around displaying alternate text (like "See details") if there's no close date (I believe this is because we _force_ a display date  - 1/1/2100). So, regarding forecasted grants that don't have a close date, would we want to show "See details" if that grant has a non-blank `close_date_explanation`? And if the forecasted grant has both a blank close date _and_ a blank `close_date_explanation`, shall we display "Not yet issued"? This text would show up in the grants table, under the Close Date column, to be clear.

**Comment by masimons** [comment]
@ClaireValdivia Thank you for talking this through with me!

Ok, that all sounds great. And, just to make sure I also understand what to do in the case that there is a _blank_ close date/estimated application due date, but a _non-blank_ close date explanation, shall I display "See details" under the Close Date column of the table? Regardless of opp status?

(In the grant details page, I will just display the full text of the close date explanation)

**Comment by ClaireValdivia** [comment]
Yes, that is correct, @masimons!
